### DISCUSSION TITLE
Selecting if-features in a non-interactive mode with `yanglint`

### DISCUSSION THREAD

**Discussion by hellt** [CESNET/libyang#2309] [Q&A] [answered]
Hi all,
first of all, thanks to all the maintainers and contributors for this amazing toolchain.

I am trying to use it in the CI pipeline to lint a big yang model set that we work on and my question is how to construct the CLI command that would enable required features.

Consider the following file layout:

```bash
_features.yang # contains features definitions
module-a.yang
module-b.yang
module-c.yang
```

I want to lint `module-c`, and `modules-c` can import, augment and contain leafrefs targeting `module-a` and `module-b`. Also all of these modules contain some `if-feature` containers.

How should I invoke `yanglint` to lint `module-c` while providing a set of features from `_features.yang` in non-interactive way?

I tried different ways, but the only thing that partilly worked is to specify `-i -i` flag to make all referenced modules implemented, but this enabled all features, and not the subset of them, as I needed.

Appreciate if someone knows how to tackle this.

**Comment by michalvasko** [comment, marked as answer]
Not sure if you are using a really old version or just missed the `-F` parameter in `yanglint -h`, but it allows for fine-tuning of all the features for all the modules.

**Reply by hellt** [reply]
using the latest, but I struggle to understand how `-F` is supposed to work
do I need to explicitly provide `_features.yang` in the command like: `yanglint _features.yang -F name:feature module-c.yang` on the CLI or if `module-c.yang` uses any features yanglint would be able to set them based on recursive imports?

**Reply by michalvasko** [reply]
You should be able to define features of any implemented modules in the context that is created. Whether a module is implemented explicitly or by an import if of no importance. But note that just because a module is imported, it does not always mean it is implemented as well so you may have to specify the module explicitly even when it is imported to implement it.

**Reply by hellt** [reply]
thanks @michalvasko 
is the notion of `implemented` vs `imported` explained somewhere for me to get a better grasp at differences?
I have been trying various things on the CLI, like specifying the `features.yang` explicitly, adding `-i -i`, and mixing the order of those, but I couldn't really tell if the feature selector selects the features I need, because there is not way to print the selected features like I can in the interactive mode.

**Reply by michalvasko** [reply]
It is explained in the [RFC](https://www.rfc-editor.org/rfc/rfc7950.html#section-5.6.5) but you should get an error if you try to enable a feature in an unimplemented module. That is also why I do not understand the reason for your issues, if you specify all the features that should be enabled, they will or you will get an error.

**Reply by hellt** [reply]
and if the features module is written as such:

```yang
module srl_nokia-features {
  yang-version 1.1;
  namespace "urn:nokia.com:srlinux:general:features";
  prefix srl_nokia-feat;

  feature admin-user-roles {
    description
      "Support for configurable roles for the admin user";
  }

  feature anycast-gw {
    description
      "Support for bridged feature and anycast-gw configuration on interface";
  }
```

is it correct to specify a list of enabled features like in the example below?

```
-F srl_nokia-features:admin-user-roles,anycast-gw
```

**Reply by michalvasko** [reply]
Based on the documentation it is, but I would use `-F "srl_nokia-features:admin-user-roles,anycast-gw"` to prevent errors with shell interpreting parts of the string.
